pkg/syncer/namespace/namespace_downstream_controller.go

```diff
@@ -100,29 +99,45 @@ func NewDownstreamController(
 			return downstreamClient.Resource(namespaceGVR).Delete(ctx, namespace, metav1.DeleteOptions{})
 		},
 		upstreamNamespaceExists: func(clusterName logicalcluster.Name, upstreamNamespaceName string) (bool, error) {
-			_, err := upstreamInformers.ForResource(namespaceGVR).Lister().ByCluster(clusterName).Get(upstreamNamespaceName)
+			lister, known, synced := ddsifForUpstreamSyncer.Lister(namespaceGVR)
+			if !known || !synced {
+				return false, errors.New("informer should have already been created and synced for namespaces in the upstream syncer informer factory")
+			}
+			_, err := lister.ByCluster(clusterName).Get(upstreamNamespaceName)
 			if apierrors.IsNotFound(err) {
 				return false, nil
 			}
-			return !apierrors.IsNotFound(err), err
+			if err != nil {
+				return false, err
+			}
+			return true, nil
 		},
 		getDownstreamNamespace: func(downstreamNamespaceName string) (runtime.Object, error) {
-			return downstreamInformers.ForResource(namespaceGVR).Lister().Get(downstreamNamespaceName)
+			lister, known, synced := ddsifForDownstream.Lister(namespaceGVR)
+			if !known || !synced {
+				return nil, errors.New("informer should have already been created and synced for namespaces in the downstream informer factory")
+			}
+			return lister.Get(downstreamNamespaceName)
 		},
 		listDownstreamNamespaces: func() ([]runtime.Object, error) {
-			return downstreamInformers.ForResource(namespaceGVR).Lister().List(labels.Everything())
+			lister, known, synced := ddsifForDownstream.Lister(namespaceGVR)
+			if !known || !synced {
+				return nil, errors.New("informer should have already been created and synced for namespaces in the downstream informer factory")
+			}
+			return lister.List(labels.Everything())
 		},
 		isDowntreamNamespaceEmpty: func(ctx context.Context, namespace string) (bool, error) {
-			gvrs, err := syncerInformers.SyncableGVRs()
-			if err != nil {
-				return false, err
+			listers, notSynced := ddsifForDownstream.Listers()
```

Comment:
elsewhere `Lister(gvr)` tells us whether it is syncer. Here it tells us what is not synced.

Reply:
Fixed in commit da89d306186c852a5f9f9ab2a3a549f896f5f7e0 by the fact that I refactored the ddsif a bit to do some cleanup, and have an `Informers()` method (instead of `Listers()`) which returns a map of `GenericLister`s so that it allows directly getting either an `SharedIndexInformer` or  a `Lister` from this map.
This allowed removing the additional `Lister()` and `Informer()` calls without loosing any feature.